### DISCUSSION TITLE
Trying to match your usage to your bill? Read this first.

### DISCUSSION THREAD

**Discussion by madpilot** [amberelectric/public-api#50] [Announcements]
Hey everyone!

I know there are a bunch of you that want to use this API to reconcile your bill. Here are a few things to consider if you are going to do that.

## We recently rolled out a new pricing engine

Over the past six months, we've been incrementally been rolling out a new pricing engine (You may have seen a notification in-app when you were enabled).

The old pricing engine suited our purpose in the beginning, but we outgrew it (It couldn't do the time of use properly for example). If you have been studiously writing down the numbers you see in the app and wonder why they now don't match the API, it's entirely possible is because you were seeing the results of the old API. This is particularly relevant if you aren't on the network's reference anytime tariff.

## The app pricing system and our billing system aren't the same things.

The system that calculates prices in the app (and the API) isn't the same as the system that generates your bill, so it's very likely there will be some small discrepancies. Ultimately, our billing system is the "source of truth", however, we use the app pricing engine to double-check the values coming out of the billing system. The app pricing engine is a clean-room implementation, so we have to do a bit of digging to find out which side is responsible for any bugs we find.

There are also a couple of rules that aren't represented in the API - for example, we have told our billing provider to never bill a negative FiT (i.e. when you owe us money for the power you generate - prices can go negative, remember?).

If you find something that doesn't seem right, email us at info@amber.com.au and we can have a look at it.

## We haven't implemented all historic tariffs. Yet.

Because the app only shows you prices back 90 days, we originally focussed on getting all 2020-21 tariffs finished first. 

For Non-Victorian customers, 2021-22 are done (Vic customers will be done at the end of ~August~ September). 

We are in the process of backfilling 2019-20 tariffs at the moment - some are done, some aren't. We _may_ do 2018-19 if there are enough people that need it, but we didn't have that many customers in those days.

if you want to know how far your particular tariff goes back, ask us and we can look it up (And we'll probably just implement it for you - they don't take long, they are just tedious).

## We can't price block tariffs properly

Given a bunch of technical reasons and that block tariffs are being phased out in favour of time of use tariffs (especially for smart meter users), we treat block tariffs like anytime tariffs, and ignore the second block price. If you regularly go over your block 1 threshold, the API will show a lower total than is actually the case.

## Computers are bad at numbers

In particular floating-point numbers. Internally we use a Big implementation to ensure things are right to the edge, but it's too hard to serialise into the API, so we live with JSON numbers. Just pretend `0.1 + 0.2 = 0.30000000000000004` and all will be well.

## We make mistakes

We are literally cutting and pasting numbers from PDFs, and we can do it wrong. There over 200 different tariffs in our system, and it's completely possible we pasted something wrong.

Again, if you see a problem - email us at info@amber.com.au and we'll investigate it.

**Comment by GrahamLea** [comment]
So, specific question on this. I'm processing my data from a January 2022 bill. When I sum up all the `kwh` from general channel usage records out of the API, I get the exact amount of kWh on our bill. Summing up `kwh * spot_per_kwh` and dividing the total by 1.1 (to remove GST from the API's price) gets... something close-ish? The total on the bill is almost exactly 5% higher than the sum from the API data. Doing the same for Controlled Load has a similar result, although it's 5.25% higher. (I round only on the end result, FWIW, not per half hour.)

Any idea if I'm falling afoul of something listed above, or is there another explanation? I'm expecting small discrepancies, but a 5% higher total price on general usage doesn't seem like the sum of rounding errors, IME. Have I missed some pricing component that's included in the bill's "Wholesale" line but isn't part of `spot_per_kwh`?

**Reply by GrahamLea** [reply]
Might as well drop in the code so you know 100% what I'm doing (in Python):
```
general_usages: List[Usage] = list(filter(lambda u: u.channel_type == ChannelType.GENERAL, usages))
invoice.general_wholesale_total_amount_cents = int(sum([u.kwh * u.spot_per_kwh for u in general_usages]) / 1.1)
invoice.general_wholesale_total_kwh = sum([u.kwh for u in general_usages])
```

**Reply by joho** [reply]
@madpilot I was checking the API docs and I don't think I can see loss factor anywhere, which is a likely candidate for this maths to not work out.

**Reply by madpilot** [reply]
Ah, good call out @joho 

You'll also need your lost factor (Multiple the spot price by 1.0499726) - that is specific to your local distribution point.

**Reply by GrahamLea** [reply]
Awesome. That's got those line items to within 1c. Thanks guys!

> that is specific to your local distribution point

local distribution point = my substation? my pole? Like it's specific to the site's location, not even the DNSP?

It would be nice to see it somewhere in the API so generic tools can make use of it.

**Reply by joho** [reply]
https://aemo.com.au/-/media/files/electricity/nem/security_and_reliability/loss_factors_and_regional_boundaries/2021-22/marginal-loss-factors-for-the-2021-22-financial-year.pdf?la=en

Transmission node (I've never actually figured out what specific physical thing that is), probably substation-ish?

**Reply by joho** [reply]
We are still working out our best strategy for sharing some of the slow-changing (but still changing) elements that input into prices and bills on the API, but intend to _somehow_ at _some point_

**Reply by madpilot** [reply]
It's made up of two numbers: The Distribution Loss Factor, which accounts for losses between the transformer and your house, and the Marginal Loss Factor which accounts for losses when transmitting power between transformers (more or less)

Just trying to understand exactly what you are doing here - the perKwh price already has all this rolled up, so I'm interested in the use case here. We can expose this stuff but I'm trying to work out where it lives.

**Reply by GrahamLea** [reply]
Sure, sorry for the lack of context. I'm writing a tool that re-creates an estimate of an invoice, line by line, with a view to being able to test out how I would have been invoiced for different tariffs for historical periods of usage. As you know, the rolled up price per period includes wholesale, tariff and non-tariff components, so the easy way for me to test different tariff components is if I build the lines up from scratch. Also, line by line is the easiest way to recognise which parts are wrong.

For my own use, I think I've now got pretty much all the numbers that will make it work.

If I want to make it useful for others to use, there's numbers in this discussion that are specific to a site or state that, at present, they'll need to fill out in config files.

FWIW, I don't think the API will ever have everything to make something like this work without any provided config. For example, I doubt you're going to start publishing all the tariff components in the API, or even which tariff each channel is on, so there will always be some data entry for people wanting to use it. (I'm not going to type all the tariffs in myself!) So putting some of this other stuff I'm asking about into the API isn't instantly going to make what I'm doing plug-and-play. I'm sure your team can figure out whether it has utility for other use cases or not.

Really appreciating the help!

**Reply by madpilot** [reply]
Ah, cool. I get what you are trying to do now.

We have this tool on our roadmap (We have an internal version already).

At this stage, we aren't going to publish tariff data like that - it's actually quite hard to model statically, more so modelling it so that it makes sense to a lay person. Time of Use tariffs can get quite fruity. There are also a bunch of network-specific exceptions.

That said, we can publish data like loss factors and tariff codes fairly easily. Let me mull it over - there are a few API improvements that are in the pipeline, so I can probably make a case for getting some time to get them sorted.

**Reply by madpilot** [reply]
I haven't included the loss factors in [today's update](https://github.com/amberelectric/public-api/discussions/100) because I want to think about the structure a little bit more. I'll let you know when it hits - should be soon.

**Reply by GrahamLea** [reply]
Looks like loss factor is just part of the tariff for the channel ([for Ausgrid](https://www.ausgrid.com.au/-/media/Documents/Regulation/Pricing/PList/AUSGRID-NETWORK-PRICE-LIST-FY2021-22.pdf), anyway), not related to specific infrastructure.

**Reply by madpilot** [reply]
That is half of it - you need the MLF too (which is linked to your TNI)

https://www.aemc.gov.au/sites/default/files/2019-07/MLF%20Methodology%20Overview%20-%20AEMO%20presentation%20Mark%20Stedwell%204%20July%202019%20rev1.pdf

**Reply by madpilot** [reply]
Multiple the two numbers and you get your loss factor

**Reply by GrahamLea** [reply]
Sheesh, they make it easy, don't they?

**Reply by joho** [reply]
Nope, and we haven't even touched on demand tariffs yet.

**Reply by GrahamLea** [reply]
Oh, I already spent hours understanding how that works. I think. 😬

**Reply by GrahamLea** [reply]
For anyone reading along, I found the 2021-22 list of MLFs online here. There isn't necessarily a straightforward way to figure out which is yours, though.
https://aemo.com.au/-/media/files/electricity/nem/security_and_reliability/loss_factors_and_regional_boundaries/2021-22/marginal-loss-factors-for-the-2021-22-financial-year.pdf?la=en

**Comment by madpilot** [comment]
Yeah, you are missing a couple of things (all ex-GST)

For NSW:

Amber hedging: 0.5 c/kWh
Environment Fee: 1.851 c/kWh
Market Fee: 0.14528 c/kWh
Carbon Offset Fee: 0.1 c/kWh

Though, this is the point of having the customer price in the API, so you don't need to know any of this...

**Reply by scotta1725** [reply]
out of curiosity, what are those for Vic?

**Reply by madpilot** [reply]
Amber hedging: 0.7 c/kWh
Environment Fee: 2.8459644 c/kWh
Market Fee: 0.2452 c/kWh
Carbon Offset Fee: 0.1 c/kWh

**Reply by GrahamLea** [reply]
Thanks. I had those as separate items, as that's how they appear on the biil. Didn't realise the rates were different by state, though, so that's good to know.

**Reply by GrahamLea** [reply]
Hmmmm... some of those rates aren't matching what's on my bill. I see (ex. GST):

Price Protection Hedging (= Amber hedging?): 1.00 c/kWh
Environmental Certificate Cost (= Environment Fee?): 2.01 c/kWh
(Market fee looks about right)
Carbon Neutral Offset 100% (= Carbon Offset Fee?): 0.22 c/kWh

**Reply by madpilot** [reply]
Sorry - I was looking at the 2021 file, not the 2021-22 file - the Environmental Certificate Cost is 2.006467c, and Carbon offset is 0.22c

Let me check that hedging fee though.

**Reply by madpilot** [reply]
Yep - you are right on the hedging too. I read the wrong number. (the Vic ones are right though)

**Comment by GrahamLea** [comment]
My next question here is around my bill's "Metering Charge".
I've got a total of 44.32c/day on my bill as one line.
From the Ausgrid tariffs line items for my channels, I see:
EA025: 8.0822 + 4.5918 for metering + 0.830 + 2.3644 for having generation tied to the grid
EA030: 0.2630 + 2.2904
Total of all 6 = 18.4218c
So there's an extra 25.8982c on the bill that isn't listed in the tariffs doc that I can see.
What am I missing?

**Reply by GrahamLea** [reply]
☝ Just wondering if you missed this question, or if you're looking into it.

**Reply by madpilot** [reply]
Definitely missed it - sorry.

My guess is the smart meter charge, which goes to the metering company (in your case Intellihub). But this is outside my specific knowledge. I'll ask the finance team.

**Reply by madpilot** [reply]
Yeah, it's smart meter charges:

4.5918c - Networt Meter Capital Costs - a charge by the network to offset the depreciation of the Basic meters they bought - you still pay this regardless of whether you have a Smartmeter, it was a compromise that AEMO allowed when they mandated Smartmeter installations because networks complained they were losing money from not being able to fully recover the costs of their old fleet.

39.726c - Smart Meter provider charge. This is what goes to Intellihub.

**Reply by GrahamLea** [reply]
Thanks.
That's interesting. I found 6 charges in [Ausgrid's tariff list](https://www.ausgrid.com.au/-/media/Documents/Regulation/Pricing/PList/AUSGRID-NETWORK-PRICE-LIST-FY2021-22.pdf) that I thought would apply to our 3 channels (2 from each of EA025, EA030 - screenshotted below - and the 2 additional generation metering charges).

But your answer suggests that only one of those 6 (the capital charge on the general/EA025 channel) is applied by Ausgrid, the other 5 ignored, and all the balance goes to IntelliHub. Is that right?

Would it be right to assume that anyone with Amber + Ausgrid has the same charge? And that there's a similar but different charge for other DNSPs, also standard across all their customers?

And I think I know the answer, but is this figure published publicly anywhere, or is asking Amber nicely basically the only way to find it out?

<img width="492" alt="image" src="https://user-images.githubusercontent.com/754403/155641777-ca792e7e-58a3-481c-a750-29b7b6e173fc.png">

**Reply by madpilot** [reply]
Sort of. They apply _per meter_ - if you are on a basic meter, there is a meter per channel.

As you have one smart meter, there are no Non-capital charges, and there is only one capital charge.

**Comment by GrahamLea** [comment]
Is there an analogue to DLF/MLF for exports? My usage totals from the API are ~1.028  x higher than the number on my last bill.

**Reply by joho** [reply]
loss factor applies in both directions, so rooftop solar exports in a corner of the grid far from the big generators are more valuable than rooftop solar close in.

**Reply by GrahamLea** [reply]
Nice, that makes sense.
Where does the rate come from? The `DLF * MLF` on my general channel is 1.0499726, but the factor on the exports appears to be around 1.028 (or 0.9728 once inverted for generation).
(I tried looking through [the AEMO DLF list](https://aemo.com.au/-/media/files/electricity/nem/security_and_reliability/loss_factors_and_regional_boundaries/2021-22/distribution-loss-factors-for-the-2021-22-financial-year.pdf?la=en) and [MLF list](https://aemo.com.au/-/media/files/electricity/nem/security_and_reliability/loss_factors_and_regional_boundaries/2020-21/marginal-loss-factors-for-the-2020-21-financial-year.pdf?la=en), but couldn't see anything about embedded generation from residential on Ausgrid, only for specific sites.)

**Reply by madpilot** [reply]
If you use the formula `0.14794 + (spot price * loss factor)` do you get closer to the right answer? There is a market charge adjustment (That works in your favour).

**Reply by GrahamLea** [reply]
Ah yes, that's done the trick!
So, I'm essentially getting _paid_ a 0.14794c/kWh "Market Charge" for embedded generation, and that doesn't have the loss factor applied to it (because loss factors are DNSP transmission things). But it's all mixed into the Solar Exports line item. 
Thanks!

**Reply by madpilot** [reply]
Yes. (For those following, the market charge is different for each state)
